### PR TITLE
[DOCS] Add example for HammingDiversityLogitsProcessor

### PR DESCRIPTION
**Pull request by jessthebp**
# What does this PR do?

Adds an example to the docstring of HammingDiversityLogitsProcessor class definition in [this file](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/logits_process.py).

Part of the docs work on [Generate: have an example on each logits processor class docstring ](https://github.com/huggingface/transformers/issues/24783)

Changes -

-   Added example
-   Added some more info about beam search and how hamming diversity works with it
-   Added warning/tip about resources

Happy to edit it all down a bit if it's too much stuff!

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests), Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.

## Who can review?

@gante

### CODE REVIEW

**Comment by jessthebp** [comment]
@gante -- hey! Currently, the checks are failing because of some changes in the docs, but not sure if I can/should fix those-- using black, the files I've changed seem to be formatted correctly? Is there something I need to change in the docs to have this merge?

**Review by gante — COMMENTED** [review]
Thank you for opening the PR! The descriptions are very thorough, I wish I had them when I first read about group beam search 🔥 

I have two requests, before we merge the PR:
1. You should be able to fix the CI errors if you run `make fixup` in your terminal, commit the changes, and then push them :)
2. Although the example is good, it would be even better if it could be done entirely with `generate` calls, as opposed to with some `group_beam_search` calls! Most users stay at a `generate` level, so an example using `generate` is much more powerful 💪

**Comment by jessthebp** [comment]
@gante Fixes made! :)

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -1085,20 +1085,134 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+
+    <Tip>
+
+    Diverse beam search can be particularly useful in scenarios where a variety of different outputs is desired, rather
+    than multiple similar sequences. It allows the model to explore different generation paths and provides a broader
+    coverage of possible outputs.
+
+    </Tip>
+
+    <Warning>
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1102]:
```suggestion
    <Tip warning={true}>
```
(this is the correct syntax for the warning block :) )

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -1085,20 +1085,134 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+
+    This logits processor can be resource-intensive, especially when using large models or long sequences.
+
+    </Warning>
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1106]:
```suggestion
    </Tip>
```

**Review by gante — COMMENTED** [review]
(the doc preview should work after correcting the syntax in the warning block)

File: examples/flax/text-classification/run_flax_glue.py [hunk 3]
```diff
@@ -212,7 +212,7 @@ def __post_init__(self):
             if self.validation_file is not None:
                 extension = self.validation_file.split(".")[-1]
                 assert extension in ["csv", "json"], "`validation_file` should be a csv or a json file."
-        self.task_name = self.task_name.lower() if type(self.task_name) == str else self.task_name
+        self.task_name = self.task_name.lower() if isinstance(self.task_name, str) else self.task_name
```
[the remark this answers is not in the document]
Comment (jessthebp) [reply examples/flax/text-classification/run_flax_glue.py]: Should all be updated + reverted

File: src/transformers/generation/logits_process.py [hunk 4]
```diff
@@ -1085,20 +1085,134 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+                    of other groups. -- This division promotes diversity by ensuring that beams within different groups
+                    explore different paths. -- For instance, if `num_beams` is 6 and `num_beam_groups` is 2, there
+                    will be 2 groups each containing 3 beams. -- The choice of `num_beam_groups` should be made
+                    considering the desired level of output diversity and the total number of beams.
+
+
+    Example: the below example shows a comparison before and after applying Hamming Diversity.
+
+    ```python
+            from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1154]: Final nit 1: let's remove indentation in the example, it shows up awkwardly in the rendered docs (see [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_25481/en/internal/generation_utils#transformers.HammingDiversityLogitsProcessor) the preview)

File: src/transformers/generation/logits_process.py [hunk 5]
```diff
@@ -1085,20 +1085,134 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+                # summary 2: the solar system formed 4.6 billion years ago from the collapse of a giant interstellar molecular cloud. of the objects that orbit the Sun directly, the largest are the eight planets, says john mccartney jr.
+                # summary 3: solar system formed 4.6 billion years ago from collapse of interstellar molecular cloud. largest of the eight planets orbit the Sun directly, with the remainder being smaller objects, such as dwarf planet and small solar System bodies - nicolaus mills-simons: the largest are the dwarf worlds and the solar systems' bodies.
+
+            print("\nNon-Diverse Summaries:")
+            for summary in summaries_non_diverse:
+                print(summary)
+                # summary 1:  the solar system formed 4.6 billion years ago from the collapse of a giant interstellar molecular cloud.
+                # summary 2:  the solar system formed 4.6 billion years ago from the collapse of a giant interstellar molecular cloud.
+    ```
+                For more details, see [Diverse Beam Search: Decoding Diverse Solutions from Neural Sequence
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1213]: Final nit 2: let's move this to above the example, and decrease the indentation -- it is not rendering correctly (see [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_25481/en/internal/generation_utils#transformers.HammingDiversityLogitsProcessor) the preview)

**Review by gante — COMMENTED** [review]
Awesome 🔥 

There are two rendering nits that should be fixed, but other than that I'm happy with the PR. Thank you for documenting this complex logits processor! 💛

**Review by gante — APPROVED** [review]

**Review by ArthurZucker — COMMENTED** [review]
Thanks for working on this!

File: src/transformers/generation/logits_process.py (hunk 4 above)
Comment (gante) [reply src/transformers/generation/logits_process.py:1154]: oh, yeah, and the missing `>>>` / `...`, good catch!

**Comment by gante** [comment]
@jesspeck thank for you iterating and thank you for the contribution 🤗
